[BUGFIX] batch.columns() returns empty list for Redshift batches

leodrivera opened the pull request:
## 🔗 Related Issues

Fixes #11532 - [batch.columns() returns empty list [] for Redshift batches, breaking all column-based validations](https://github.com/great-expectations/great_expectations/issues/11532)

Fixes #11181 - [Problems on retrieving batch when using redshift_datasource](https://github.com/great-expectations/great_expectations/issues/11181)

## 🐛 Problem

When using Great Expectations with Amazon Redshift, the `batch.columns()` method returns an empty list `[]` even when the batch contains data. This causes **all expectations that reference column names to fail** with `InvalidMetricAccessorDomainKwargsKeyError`.

### Root Causes

1. **`information_schema.columns` returns empty**: Some Redshift clusters have restricted access to `information_schema.columns`, causing queries to return 0 rows even when tables exist and have columns.

2. **Schema/Table parsing issue**: Many GX Redshift configurations use `table_name="schema.table"` with `schema_name=None`. The original code didn't properly split this before querying `information_schema`.

3. **SQLAlchemy 2.x compatibility**: The code attempted to use `len()` on a `CursorResult` object, which doesn't support `len()` in SQLAlchemy 2.x.

## ✅ Solution

Implemented a robust fallback mechanism in `RedshiftExecutionEngine.ColumnTypes._sqlalchemy()`:

1. **Query `information_schema.columns` first** (optimized path)
2. **If it returns empty**, fall back to `SELECT * FROM schema.table LIMIT 0`
3. **Extract column names** from the query result using `result.keys()`
4. **Handle SQLAlchemy 2.x** by converting `CursorResult` to list before operations

### Key Changes

#### `execution_engine/redshift_execution_engine.py`

- Added fallback to `SELECT * LIMIT 0` when `information_schema` returns empty
- Fixed SQLAlchemy 2.x compatibility by converting `CursorResult` to list before operations
- Implemented `_get_table_schema()` method to properly parse `table_name="schema.table"` format into separate schema and table components
- Added comprehensive logging (debug/info/warning) for troubleshooting column retrieval

#### `expectations/metrics/table_metrics/table_column_types.py`

- Enhanced Redshift-specific handling in `_get_sqlalchemy_column_metadata` function
- Added early return path for Redshift that normalizes `table_name="schema.table"` into separate schema and table components before calling `get_sqlalchemy_column_metadata`
- Added comprehensive debug logging to track schema/table parsing and column metadata retrieval

#### `expectations/metrics/util.py`

- Added `schema_name` parameter to `column_reflection_fallback` function signature (defaults to `None`)
- For Redshift dialect, properly composes `schema.table` reference when executing `SELECT * LIMIT 0` fallback query
- Uses `sa.text()` wrapper for Redshift table references to handle fully-qualified table names correctly

## 🔄 Compatibility

- ✅ Redshift (all clusters, including those with restricted `information_schema`)
- ✅ Configurations with `table_name="schema.table"` and `schema_name=None`
- ✅ Configurations with `table_name="table"` and `schema_name="schema"` (standard)

leodrivera committed:
fix(redshift): add fallback for empty batch.columns() on Redshift

Add fallback mechanism using SELECT * LIMIT 0 when information_schema
returns empty, fixing batch.columns() returning empty list for Redshift
batches. Also fixes SQLAlchemy 2.x compatibility and schema/table parsing.

Fixes #11532

Fixes #11181

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

leodrivera committed:
fix: resolve ruff linter errors in Redshift execution engine

- Replace dict.keys() with direct dict iteration (SIM118)
- Use logging.exception instead of logging.error (TRY400)
- Fix line length violations (E501)
- Replace non-breaking hyphen with regular hyphen in comment (RUF003)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

leodrivera committed:
fix(redshift): correct column name retrieval in fallback query

    Update the fallback mechanism in the Redshift execution engine to correctly retrieve column names from the result of
the fallback query. This change ensures that the column names are accurately extracted, improving the reliability of the batch.columns() functionality.

tyler-hoffman commented on great_expectations/expectations/metrics/util.py:
```diff
@@ -461,6 +473,7 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915 # FIXME CoP
     selectable: sqlalchemy.Select,
     dialect: sqlalchemy.Dialect,
     sqlalchemy_engine: sqlalchemy.Engine,
+    schema_name: Optional[str] = None,
```
This looks to be unused

tyler-hoffman reviewed (commented):
Thanks @leodrivera ! I just kicked off the test suite to make sure everything still passes. The implementation looks solid to me! Would you mind taking a look at `tests/integration/data_sources_and_expectations/data_sources/test_redshift.py` to see if you can add a test to ensure we don't introduce regressions with this later?

leodrivera committed:
refactor: remove unused schema_name parameter from column metadata functions

This change eliminates the schema_name parameter from the get_sqlalchemy_column_metadata and column_reflection_fallback functions, as it was not being utilized. This cleanup enhances code clarity and maintainability.

leodrivera committed:
feat(redshift): enhance column type handling and add tests for various data types

This update improves the Redshift execution engine by implementing parameterized queries for retrieving column metadata, enhancing security against SQL injection. It also introduces a fallback mechanism for custom SQL queries and expands the test suite to cover various Redshift data types, ensuring accurate type validation in batch processing.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

leodrivera commented:
> Thanks @leodrivera ! I just kicked off the test suite to make sure everything still passes. The implementation looks solid to me! Would you mind taking a look at `tests/integration/data_sources_and_expectations/data_sources/test_redshift.py` to see if you can add a test to ensure we don't introduce regressions with this later?

Sure

I've added some tests and also make some small improvements like enhancing security against SQL injection.

leodrivera committed:
feat(redshift): improve column type retrieval and fallback mechanisms

This update enhances the Redshift execution engine by implementing a more robust method for querying column types from information_schema, including a fallback to SELECT * LIMIT 0 when necessary. The changes also introduce a new method for processing information_schema rows and improve the handling of custom SQL queries, ensuring better reliability and security in column metadata retrieval.

leodrivera committed:
test(redshift): add test for non-empty batch.columns() functionality

This commit introduces a new test to ensure that the batch.columns() method correctly returns column names for Redshift batches, validating the recent fix that prevents it from returning an empty list. The test checks for the presence and correctness of expected column names, enhancing the reliability of the Redshift execution engine.

leodrivera commented:
Thanks

leodrivera committed:
refactor: simplify type hinting for table_selectable in column metadata retrieval

This commit removes the redundant type hint for the table_selectable variable in the _get_sqlalchemy_column_metadata function, enhancing code clarity without altering functionality.

leodrivera commented:
The static-analysis check failed. I've added a new commit to adjust it.

NathanFarmer committed:
Merge branch 'develop' into fix/redshift-table-column-types

NathanFarmer committed:
Merge branch 'develop' into fix/redshift-table-column-types

leodrivera committed:
Merge branch 'develop' into fix/redshift-table-column-types

leodrivera committed:
Merge branch 'develop' into fix/redshift-table-column-types

leodrivera commented:
I left this as a draft because I wasn’t able to review it properly at the end of the year.

@NathanFarmer
Unfortunately, the ResourceWarning is still occurring in the `test_sqlite_single_column_complete_result_format_id_pk` test. Let me know if there’s anything I can do to help with the investigation or fix.

NathanFarmer commented:
Thanks @leodrivera! This fix will be included in the next GX Core release.
